crates/walrus-service/src/node/storage/shard.rs

```diff
@@ -740,10 +824,23 @@ impl ShardStorage {
         self.recovery_any_missing_slivers(node, config, epoch)
             .await?;
 
-        let mut batch = self.shard_status.batch();
-        batch.insert_batch(&self.shard_status, [((), ShardStatus::Active)])?;
-        batch.delete_batch(&self.shard_sync_progress, [()])?;
-        batch.write()?;
+        self.shard_status.write().await.with_shard_status_mut(
+            |shard_status| -> Result<(), TypedStoreError> {
+                let mut batch = self.shard_sync_progress.batch();
+                let current_status = shard_status.get(&())?;
+                // If the shard is in active sync or active recover, we need to update the status to
+                // active. Otherwise, the shard status is interrupted by another process and are in
+                // unexpected state and we should not update the status.
+                if current_status == Some(ShardStatus::ActiveSync)
+                    || current_status == Some(ShardStatus::ActiveRecover)
+                {
+                    batch.insert_batch(shard_status, [((), ShardStatus::Active)])?;
+                }
```

Comment:
Question: Just to make sure I fully understand the fix: Is this the core of the solution as mentioned in the PR description?

> The fix for this issue is to not allow shard status to be transition to other status if it is in LockedToMove, unless, the node starts a new shard sync that explicitly requests a new shard sync task. This prevents the shard mistakenly move from LockedToMove to Active status.

Reply:
Yes, this is part of the core to address the test failure that is manifested. The real fix is all the places where with_shard_status_mut() is used: shard status transition needs to be transactional and we cannot blindly apply new status. E.g., previously we transition shards from ActiveSync to ActiveRecover, but when doing this transition, the shard may not in ActiveSync anymore.